File: javascript-matrix/shape.js
import { NeuralNetwork } from './neural.js';

const network = new NeuralNetwork({
    activation: 'sigmoid',
    layers: [ 9, 6, 3 ]
});

// Training
console.log('Training...')
const trainingCycles = network.train([
    {
        input: [
            1, 1, 1,
            1, 0, 1,
            1, 1, 1
        ],
        output: 'O'
    },
    {
        input: [
            0, 1, 0,
            1, 0, 1,
            0, 1, 0
        ],
        output: 'O'
    },
    {
        input: [
            0, 1, 0,
            1, 1, 1,
            0, 1, 0
        ],
        output: 'X'
    },
    {
        input: [
            1, 0, 1,
            0, 1, 0,
            1, 0, 1
        ],
        output: 'X'
    },
    {
        input: [
            1, 0, 0,
            0, 0, 0,
            0, 0, 0
        ],
        output: '*'
    },
    {
        input: [
            0, 1, 0,
            0, 0, 0,
            0, 0, 0
        ],
        output: '*'
    },
    {
        input: [
            0, 0, 1,
            0, 0, 0,
            0, 0, 0
        ],
        output: '*'
    }
], 0.01);
console.log(`Training done in ${trainingCycles} cycles`);
console.log(network.symbols);

// Testing
const testingData = [
    {
        input: [
            0, 1, 1,
            1, 0, 1,
            1, 1, 0
        ],
        output: 'O'
    },
    {
        input: [
            1, 0, 1,
            1, 0, 1,
            1, 1, 0
        ],
        output: 'O'
    },
    {
        input: [
            1, 0, 0,
            1, 1, 1,
            0, 0, 1
        ],
        output: 'X'
    },
    {
        input: [
            0, 0, 1,
            1, 1, 1,
            1, 0, 0
        ],
        output: 'X'
    },
    {
        input: [
            1, 0, 1,
            0, 0, 0,
            1, 0, 1
        ],
        output: 'O'
    },
    {
        input: [
            0, 1, 0,
            1, 0, 1,
            0, 1, 0
        ],
        output: 'O'
    },
    {
        input: [
            0, 0, 0,
            1, 1, 1,
            0, 0, 0
        ],
        output: 'X'
    },
    {
        input: [
            1, 0, 0,
            1, 1, 0,
            1, 0, 0
        ],
        output: 'X'
    },
    {
        input: [
            1, 0, 0,
            0, 0, 0,
            0, 0, 0
        ],
        output: '*'
    },
    {
        input: [
            0, 1, 1,
            0, 0, 0,
            0, 0, 0
        ],
        output: '*'
    }
];
for (const item of testingData) {
    const result = network.likely(item.input);
    console.log(result, item.output, result == item.output ? 'PASSED' : 'FAILED');
}

File: python-matrix/shape.py
import data
from neural import NeuralNetwork

if __name__ == '__main__':
    # Random network
    randomNetwork = NeuralNetwork(
        activation='sigmoid',
        training='random',
        layers=[ data.inputDim, 6, 6, data.outputDim ]
    )

    print('Random network training...')
    trainingCycles = randomNetwork.train(data.trainingSet, data.maxError)
    print('Training done in', trainingCycles, 'cycles!')
    print(randomNetwork.symbols)

    for item in data.testSet:
        result = randomNetwork.likely(item[0])
        print(result, item[1], result == item[1] and 'PASSED' or 'FAILED')

    # Math network
    mathNetwork = NeuralNetwork(
        activation='sigmoid',
        training='math',
        layers=[ data.inputDim, 6, 6, data.outputDim ]
    )

    print('\nMath network training...')
    trainingCycles = mathNetwork.train(data.trainingSet, data.maxError)
    print('Training done in', trainingCycles, 'cycles!')
    print(mathNetwork.symbols)

    for item in data.testSet:
        result = mathNetwork.likely(item[0])
        print(result, item[1], result == item[1] and 'PASSED' or 'FAILED')
